File: ukpsummarizer-be/cplex/cpoptimizer/examples/src/cpp/sched_rcpspmm.cpp
// -------------------------------------------------------------- -*- C++ -*-
// File: ./examples/src/cpp/sched_rcpspmm.cpp
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
//
// 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5725-A06 5725-A29
// Copyright IBM Corporation 1990, 2014. All Rights Reserved.
//
// Note to U.S. Government Users Restricted Rights:
// Use, duplication or disclosure restricted by GSA ADP Schedule
// Contract with IBM Corp.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

The MMRCPSP (Multi-Mode Resource-Constrained Project Scheduling
Problem) is a generalization of the Resource-Constrained Project
Scheduling problem (see sched_rcpsp.cpp). In the MMRCPSP, each
activity can be performed in one out of several modes. Each mode of an
activity represents an alternative way of combining different levels
of resource requirements with a related duration. Renewable and
no-renewable resources are distinguished. While renewable resources
have a limited instantaneous availability such as manpower and
machines, non renewable resources are limited for the entire project,
allowing to model, e.g., a budget for the project.  The objective is
to find a mode and a start time for each activity such that the
schedule is makespan minimal and feasible with regard to the
precedence and resource constraints.

------------------------------------------------------------ */

#include <ilcp/cp.h>

class FileError: public IloException {
public:
  FileError() : IloException("Cannot open data file") {}
};

int main(int argc, const char* argv[]){
  IloEnv env;
  try {
    const char* filename = "../../../examples/data/rcpspmm_default.data";
    IloInt failLimit = 30000;
    if (argc > 1)
      filename = argv[1];
    if (argc > 2)
      failLimit = atoi(argv[2]);
    std::ifstream file(filename);
    if (!file){
      env.out() << "usage: " << argv[0] << " <file> <failLimit>" << std::endl;
      throw FileError();
    }

    IloInt i, j, k;
    IloModel model(env);
    IloInt nbTasks, nbRenewable, nbNonRenewable;
    file >> nbTasks >> nbRenewable >> nbNonRenewable;
    IloCumulFunctionExprArray renewables(env, nbRenewable);
    IloIntExprArray nonRenewables(env, nbNonRenewable);
    IloIntArray capRenewables   (env, nbRenewable);
    IloIntArray capNonRenewables(env, nbNonRenewable);
    for (j=0; j<nbRenewable; j++) {
      renewables[j] = IloCumulFunctionExpr(env);
      file >> capRenewables[j];
    }
    for (j=0; j<nbNonRenewable; j++){
      nonRenewables[j] = IloIntExpr(env);
      file >> capNonRenewables[j];
    }

    IloIntervalVarArray  tasks(env, nbTasks);
    IloIntervalVarArray2 modes(env, nbTasks);
    for (i=0; i<nbTasks; i++) {
      tasks[i] = IloIntervalVar(env);
      modes[i] = IloIntervalVarArray(env);
    }
    IloIntExprArray ends(env);
    for (i=0; i<nbTasks; i++) {
      IloIntervalVar task = tasks[i];
      IloInt d, nbModes, nbSucc;
      file >> d >> nbModes >> nbSucc;
      for (k=0; k<nbModes; ++k) {
        IloIntervalVar alt(env);
        alt.setOptional();
        modes[i].add(alt);
      }
      model.add(IloAlternative(env, task, modes[i]));
      ends.add(IloEndOf(task));
      for (IloInt s=0; s<nbSucc; ++s) {
        IloInt succ;
        file >> succ;
        model.add(IloEndBeforeStart(env, task, tasks[succ]));
      }
    }
    for (i=0; i<nbTasks; i++) {
      IloIntervalVarArray imodes = modes[i];
      for (k=0; k<imodes.getSize(); ++k) {
        IloInt taskId, modeId, d;
        file >> taskId >> modeId >> d;
        imodes[k].setSizeMin(d);
        imodes[k].setSizeMax(d);
        IloInt q;
        for (j = 0; j < nbRenewable; j++) {
          file >> q;
          if (0 < q) {
            renewables[j] += IloPulse(imodes[k], q);
          }
        }
        for (j = 0; j < nbNonRenewable; j++) {
          file >> q;
          if (0 < q) {
            nonRenewables[j] += q * IloPresenceOf(env, imodes[k]);
          }
        }
      }
    }

    for (j = 0; j < nbRenewable; j++) {
      model.add(renewables[j] <= capRenewables[j]);
    }

    for (j = 0; j < nbNonRenewable; j++) {
      model.add(nonRenewables[j] <= capNonRenewables[j]);
    }

    IloObjective objective = IloMinimize(env,IloMax(ends));
    model.add(objective);

    IloCP cp(model);
    cp.setParameter(IloCP::FailLimit, failLimit);
    cp.out() << "Instance \t: " << filename << std::endl;
    if (cp.solve()) {
      cp.out() << "Makespan \t: " << cp.getObjValue() << std::endl;
    } else {
      cp.out() << "No solution found."  << std::endl;
    }
  } catch(IloException& e){
    env.out() << " ERROR: " << e << std::endl;
  }
  env.end();
  return 0;
}

File: ukpsummarizer-be/cplex/cpoptimizer/examples/src/java/SchedRCPSPMM.java
// ---------------------------------------------------------------*- Java -*-
// File: ./examples/src/java/SchedRCPSPMM.java
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
//
// 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5725-A06 5725-A29
// Copyright IBM Corporation 1990, 2017. All Rights Reserved.
//
// Note to U.S. Government Users Restricted Rights:
// Use, duplication or disclosure restricted by GSA ADP Schedule
// Contract with IBM Corp.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

The MMRCPSP (Multi-Mode Resource-Constrained Project Scheduling
Problem) is a generalization of the Resource-Constrained Project
Scheduling problem (see SchedRCPSP.java). In the MMRCPSP, each
activity can be performed in one out of several modes. Each mode of an
activity represents an alternative way of combining different levels
of resource requirements with a related duration. Renewable and
no-renewable resources are distinguished. While renewable resources
have a limited instantaneous availability such as manpower and
machines, non renewable resources are limited for the entire project,
allowing to model, e.g., a budget for the project.  The objective is
to find a mode and a start time for each activity such that the
schedule is makespan minimal and feasible with regard to the
precedence and resource constraints.

------------------------------------------------------------ */

import ilog.concert.*;
import ilog.cp.*;
import java.io.*;
import java.util.*;

public class SchedRCPSPMM {

    private static class DataReader {

        private StreamTokenizer st;

        public DataReader(String filename) throws IOException {
            FileInputStream fstream = new FileInputStream(filename);
            Reader r = new BufferedReader(new InputStreamReader(fstream));
            st = new StreamTokenizer(r);
        }

        public int next() throws IOException {
            st.nextToken();
            return (int) st.nval;
        }
    }

    static class IntervalVarList extends ArrayList<IloIntervalVar> {
        public IloIntervalVar[] toArray() {
            return (IloIntervalVar[]) this.toArray(new IloIntervalVar[this.size()]);
        }
    }

    static IloIntExpr[] arrayFromList(List<IloIntExpr> list) {
        return (IloIntExpr[]) list.toArray(new IloIntExpr[list.size()]);
    }

    public static void main(String[] args) throws IOException {

        String filename = "../../../examples/data/rcpspmm_default.data";
        int failLimit = 30000;
        int nbTasks, nbRenewable, nbNonRenewable;

        if (args.length > 0)
            filename = args[0];
        if (args.length > 1)
            failLimit = Integer.parseInt(args[1]);

        IloCP cp = new IloCP();
        DataReader data = new DataReader(filename);
        try {
            nbTasks = data.next();
            nbRenewable = data.next();
            nbNonRenewable = data.next();
            IloCumulFunctionExpr[] renewables = new IloCumulFunctionExpr[nbRenewable];
            IloIntExpr[] nonRenewables = new IloIntExpr[nbNonRenewable];
            int[] capRenewables = new int[nbRenewable];
            int[] capNonRenewables = new int[nbNonRenewable];
            for (int j = 0; j < nbRenewable; j++) {
                renewables[j] = cp.cumulFunctionExpr();
                capRenewables[j] = data.next();
            }
            for (int j = 0; j < nbNonRenewable; j++) {
                nonRenewables[j] = cp.intExpr();
                capNonRenewables[j] = data.next();
            }

            IloIntervalVar[] tasks = new IloIntervalVar[nbTasks];
            IntervalVarList[] modes = new IntervalVarList[nbTasks];
            for (int i = 0; i < nbTasks; i++) {
                tasks[i] = cp.intervalVar();
                modes[i] = new IntervalVarList();
            }
            List<IloIntExpr> ends = new ArrayList<IloIntExpr>();
            for (int i = 0; i < nbTasks; i++) {
                IloIntervalVar task = tasks[i];
                int d = data.next();
                int nbModes = data.next();
                int nbSucc = data.next();
                for (int k = 0; k < nbModes; k++) {
                    IloIntervalVar alt = cp.intervalVar();
                    alt.setOptional();
                    modes[i].add(alt);
                }
                cp.add(cp.alternative(task, modes[i].toArray()));
                ends.add(cp.endOf(task));
                for (int s = 0; s < nbSucc; s++) {
                    int succ = data.next();
                    cp.add(cp.endBeforeStart(task, tasks[succ]));
                }
            }
            for (int i = 0; i < nbTasks; i++) {
                IloIntervalVar task = tasks[i];
                IntervalVarList imodes = modes[i];
                for(int k=0; k < imodes.size(); k++) {
                    int taskId = data.next();
                    int modeId = data.next();
                    int d = data.next();
                    imodes.get(k).setSizeMin(d);
                    imodes.get(k).setSizeMax(d);
                    int q;
                    for (int j = 0; j < nbRenewable; j++) {
                        q = data.next();
                        if (0 < q) {
                            renewables[j] = cp.sum(renewables[j], cp.pulse(imodes.get(k), q));
                        }
                    }
                    for (int j = 0; j < nbNonRenewable; j++) {
                        q = data.next();
                        if (0 < q) {
                            nonRenewables[j] = cp.sum(nonRenewables[j], cp.prod(q, cp.presenceOf(imodes.get(k))));
                        }
                    }
                }
            }

            for (int j = 0; j < nbRenewable; j++) {
                cp.add(cp.le(renewables[j], capRenewables[j]));
            }

            for (int j = 0; j < nbRenewable; j++) {
                cp.add(cp.le(nonRenewables[j], capNonRenewables[j]));
            }

            IloObjective objective = cp.minimize(cp.max(arrayFromList(ends)));
            cp.add(objective);

            cp.setParameter(IloCP.IntParam.FailLimit, failLimit);
            System.out.println("Instance \t: " + filename);
            if (cp.solve()) {
                System.out.println("Makespan \t: " + cp.getObjValue());
            }
            else {
                System.out.println("No solution found.");
            }
        } catch (IloException e) {
            System.err.println("Error: " + e);
        }
    }



}
